### PR TITLE
Add Unreal CPP & BP SDK support for Local Debugging Azure Functions

### PR DESCRIPTION
**Pull request by nima-ap-zz**
Add support in SDKGenerator to generate Unreal C++ & Blueprints SDK that supports local debugging
Also supports partial debugging.

### CODE REVIEW

File: targets/unity-v2/make.js [hunk 1]
```diff
@@ -484,7 +484,7 @@ function getCustomApiLogic(tabbing, apiCall)
 {
     if (apiCall.name === "ExecuteFunction")
         return tabbing + "var localApiServerString = PlayFabSettings.LocalApiServer;\n" +
-            tabbing + "if (!string.IsNullOrEmpty(localApiServerString))\n" +
+            tabbing + "if (!string.IsNullOrEmpty(localApiServerString) && PlayFab.Internal.PlayFabUtil.ShouldDebugFunction(request.FunctionName))\n" +
```
Comment (MGudgin) [inline targets/unity-v2/make.js]: This change should be in a different PR for Unity, yes?

File: targets/unity-v2/source/Shared/Internal/Util.cs [hunk 2]
```diff
@@ -161,6 +161,12 @@ internal static string GetLocalSettingsFileProperty(string propertyKey)
             }
             return string.Empty;
         }
+
+        internal static bool ShouldDebugFunction(string functionName)
```
Comment (MGudgin) [inline targets/unity-v2/source/Shared/Internal/Util.cs]: This change should be in a different PR for Unity, yes?

File: targets/UnrealMarketplacePlugin/makeCpp.js [hunk 3]
```diff
@@ -660,6 +661,21 @@ function getRequestActions(tabbing, apiCall, isInstanceApi) {
     return "";
 }
 
+function getCustomApiLogic(tabbing, api, apiCall, isInstanceApi) {
+    if (apiCall.name === "ExecuteFunction")
+        return tabbing + "FString localApiServer = PlayFabSettings::GetLocalApiServer();\n"
+            + tabbing + "TArray<FString> debugFunctions = PlayFabSettings::GetDebugFunctions();\n"
+            + tabbing + "if (!localApiServer.IsEmpty() && (debugFunctions.Contains(request.FunctionName) || debugFunctions.Contains(TEXT(\"*\"))))\n"
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/makeCpp.js]: I think the last clause should be use Equals, not Contains.

Comment (MGudgin) [reply targets/UnrealMarketplacePlugin/makeCpp.js]: And I'm still not certain we should do the function-based settings.

File: targets/UnrealMarketplacePlugin/makeBp.js [hunk 4]
```diff
@@ -296,3 +298,30 @@ function getAuthBools(tabbing, apiCall) {
...
     return output;
 }
+
+function getCustomApiAssignmentLogic(tabbing, apiCall) {
+    if (apiCall.name === "ExecuteFunction")
+    {
+        return tabbing + "// Check for local debugging\n"
+            + tabbing + "FString localApiServer = PlayFabCommon::PlayFabCommonUtils::GetLocalSettingsFileProperty(TEXT(\"LocalApiServer\"));\n"
+            + tabbing + "TArray<FString> debugFunctions = PlayFabCommon::PlayFabCommonUtils::GetLocalSettingsFileArrayProperty(TEXT(\"DebugFunctions\"));\n"
+            + tabbing + "if (!localApiServer.IsEmpty() && (debugFunctions.Contains(TEXT(\"*\")) || debugFunctions.Contains(request.FunctionName)))\n"
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/makeBp.js]: Your clauses here are not in the same order as the same ones in makeCpp.js. Also see comments below about Equals vs Contains and whether we should even do the name based settings

File: targets/UnrealMarketplacePlugin/makeBp.js [hunk 5]
```diff
@@ -296,3 +298,30 @@ function getAuthBools(tabbing, apiCall) {
...
+            + tabbing + "{\n"
+            + tabbing + "    FString endpoint = TEXT(\"/CloudScript/ExecuteFunction\");\n"
+            + tabbing + "    endpoint.RemoveFromStart(TEXT(\"/\"));\n"
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/makeBp.js:310]: Why not just remove the leading / from the string on the line above?

File: targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs [hunk 6]
```diff
@@ -0,0 +1,56 @@
+﻿<%- copyright %>
+
+#include "PlayFabCommonUtils.h"
+
+using namespace PlayFabCommon;
+
+FString PlayFabCommonUtils::GetTempDir()
+{
+    // Unix
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs]: Which one works for Mac?

File: targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    FString localSettingsFileContent = "";
+    return FFileHelper::LoadFileToString(localSettingsFileContent, *localSettingsFilePath)
+        ? localSettingsFileContent
+        : FString();
+}
+
+TSharedPtr<FJsonObject> PlayFabCommonUtils::GetLocalSettingsFileJson()
+{
+    FString fileContent = PlayFabCommonUtils::GetLocalSettingsFileContent();
+    TSharedPtr<FJsonObject> JsonObject = MakeShareable(new FJsonObject());
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs]: nit: lower case j

File: targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    TSharedRef<TJsonReader<>> JsonReader = TJsonReaderFactory<>::Create(fileContent);
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs]: nit: Lower case j

File: targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+{
+    FString tempDirPath = PlayFabCommonUtils::GetTempDir();
+    FString localSettingsFilePath = FPaths::Combine(tempDirPath, TEXT("playfab.local.settings.json"));
+    FString localSettingsFileContent = "";
+    return FFileHelper::LoadFileToString(localSettingsFileContent, *localSettingsFilePath)
+        ? localSettingsFileContent
+        : FString();
+}
+
+TSharedPtr<FJsonObject> PlayFabCommonUtils::GetLocalSettingsFileJson()
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs:31]: Why is this a SharedPtr?

File: targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+{
+    FString fileContent = PlayFabCommonUtils::GetLocalSettingsFileContent();
+    TSharedPtr<FJsonObject> JsonObject = MakeShareable(new FJsonObject());
+    TSharedRef<TJsonReader<>> JsonReader = TJsonReaderFactory<>::Create(fileContent);
+    return (FJsonSerializer::Deserialize(JsonReader, JsonObject) && JsonObject.IsValid())
+        ? JsonObject
+        : MakeShareable(new FJsonObject);
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs]: This looks weird without the () on the ctor call

File: targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs [hunk 11]
```diff
@@ -0,0 +1,56 @@
+﻿<%- copyright %>
+
+#include "PlayFabCommonUtils.h"
+
+using namespace PlayFabCommon;
+
+FString PlayFabCommonUtils::GetTempDir()
+{
+    // Unix
+    FString tempDir = FPlatformMisc::GetEnvironmentVariable(TEXT("TEMPDIR"));
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs]: I'd be tempted to shorten this code by having an array of Environment variables names that I iterated over.

File: targets/UnrealMarketplacePlugin/templates/PlayFab/PlayFab_API.h.ejs [hunk 12]
```diff
@@ -93,12 +93,16 @@ public:
 
     /** PlayFab Request Info */
     FString PlayFabRequestURL;
+<% if (api.name === "CloudScript") {
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/templates/PlayFab/PlayFab_API.h.ejs:96]: This looks funky. Should the %> on line 97 really be at the end of line 96?

File: targets/UnrealMarketplacePlugin/makeCpp.js (hunk 3 above)
Comment (nima-ap-zz) [reply targets/UnrealMarketplacePlugin/makeCpp.js]: Well if we do equals it would be more code cause then we'd have to make sure the size is 1 and that the first element is *, while to me it would make more sense to say * umbrellas everything else.

Comment (nima-ap-zz) [reply targets/UnrealMarketplacePlugin/makeCpp.js]: What are you referring to in terms of function based settings?

File: targets/UnrealMarketplacePlugin/makeBp.js (hunk 5 above)
Comment (nima-ap-zz) [reply targets/UnrealMarketplacePlugin/makeBp.js:310]: Oh, the text shouldn't be in there. It should be `api.path` or whatever it is that has that value hence why I remove the / from the start. Will fix to remove the hardcoded string.

File: targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs (hunk 6 above)
Comment (nima-ap-zz) [reply targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs]: Mac is unix no?

File: targets/UnrealMarketplacePlugin/templates/PlayFab/PlayFab_API.h.ejs (hunk 12 above)
Comment (nima-ap-zz) [reply targets/UnrealMarketplacePlugin/templates/PlayFab/PlayFab_API.h.ejs:96]: For some reason, everywhere else in the SDK that's the pattern they use for if statements and hence why I followed.

File: targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs (hunk 9 above)
Comment (nima-ap-zz) [reply targets/UnrealMarketplacePlugin/source/PlayFab/Source/PlayFabCommon/Private/PlayFabCommmonUtils.cpp.ejs:31]: That is the pattern I saw them use with creating json objects out of a file.

File: targets/UnrealMarketplacePlugin/makeCpp.js (hunk 3 above)
Comment (MGudgin) [reply targets/UnrealMarketplacePlugin/makeCpp.js]: See my comment on https://github.com/nima-ap/SDKGenerator/pull/2

**Comment by nima-ap-zz** [comment]
@MGudgin  Addressed everything. Regarding Partial debugging, resolved the comment for now but we should talk about it when you get a chance.

File: targets/UnrealMarketplacePlugin/makeBp.js [hunk 13]
```diff
@@ -307,7 +307,7 @@ function getCustomApiAssignmentLogic(tabbing, apiCall) {
             + tabbing + "TArray<FString> debugFunctions = PlayFabCommon::PlayFabCommonUtils::GetLocalSettingsFileArrayProperty(TEXT(\"DebugFunctions\"));\n"
             + tabbing + "if (!localApiServer.IsEmpty() && (debugFunctions.Contains(TEXT(\"*\")) || debugFunctions.Contains(request.FunctionName)))\n"
             + tabbing + "{\n"
-            + tabbing + "    FString endpoint = TEXT(\"/CloudScript/ExecuteFunction\");\n"
+            + tabbing + "    FString endpoint = TEXT(\"" + apiCall.url + "\");\n"
```
Comment (MGudgin) [inline targets/UnrealMarketplacePlugin/makeBp.js:309]: Wow, this line is really hard to parse!

File: targets/unity-v2/make.js [hunk 14]
```diff
@@ -484,7 +484,7 @@ function getCustomApiLogic(tabbing, apiCall)
 {
     if (apiCall.name === "ExecuteFunction")
         return tabbing + "var localApiServerString = PlayFabSettings.LocalApiServer;\n" +
-            tabbing + "if (!string.IsNullOrEmpty(localApiServerString) && PlayFab.Internal.PlayFabUtil.ShouldDebugFunction(request.FunctionName))\n" +
+            tabbing + "if (!string.IsNullOrEmpty(localApiServerString))\n" +
```
Comment (MGudgin) [inline targets/unity-v2/make.js]: I'm still confused at to why changes under targets/unity-v2 are showing up in a PR for Unreal.

**Review by MGudgin — CHANGES_REQUESTED** [review]

File: targets/unity-v2/make.js (hunk 14 above)
Comment (nima-ap-zz) [reply targets/unity-v2/make.js]: Please look at changes of the PR as a whole with respect to master instead of changes to previous PR commit when evaluating this.
Unity local debugging is already checked in; this branch contains those changes as it branched off of master post GDC which contained local debugging changes needed to SDKGenerator. 
Having clicked "Files Changes" on the PR, and Ctrl+F'd for "unity" I do not see anything come back = no changes in this PR to unity SDK.

File: targets/UnrealMarketplacePlugin/makeBp.js (hunk 13 above)
Comment (nima-ap-zz) [reply targets/UnrealMarketplacePlugin/makeBp.js:309]: I don't follow what you are asking me to do here. Can you please explain?

Comment (MGudgin) [reply targets/UnrealMarketplacePlugin/makeBp.js:309]: I don't think there is anything you can do. Just the combination of the TEXT macro and the content of that macro being generated makes the escaping of the various " characters difficult to read.

Comment (nima-ap-zz) [reply targets/UnrealMarketplacePlugin/makeBp.js:309]: In that case if every concern on this PR is addressed can we please approve and merge it so we can proceed with handing it to the SDK team for review? I don't suspect them to approve it immediately and so more work is awaiting me waiting on this to get approved.

**Review by MGudgin — APPROVED** [review]
